// @walletconnect/client/dist/cjs/controllers/expirer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Expirer = void 0;
const tslib_1 = require("tslib");
const heartbeat_1 = require("@walletconnect/heartbeat");
const logger_1 = require("@walletconnect/logger");
const time_1 = require("@walletconnect/time");
const types_1 = require("@walletconnect/types");
const utils_1 = require("@walletconnect/utils");
const events_1 = require("events");
const constants_1 = require("../constants");
class Expirer extends types_1.IExpirer {
    constructor(core, logger) {
        super(core, logger);
        this.core = core;
        this.logger = logger;
        this.expirations = new Map();
        this.events = new events_1.EventEmitter();
        this.name = constants_1.EXPIRER_CONTEXT;
        this.version = constants_1.EXPIRER_STORAGE_VERSION;
        this.cached = [];
        this.initialized = false;
        this.storagePrefix = constants_1.CLIENT_STORAGE_PREFIX;
        this.init = () => tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.initialized) {
                this.logger.trace(`Initialized`);
                yield this.restore();
                this.cached.forEach(expiration => this.expirations.set(expiration.topic, expiration));
                this.cached = [];
                this.registerEventListeners();
                this.initialized = true;
            }
        });
        this.has = topic => {
            try {
                const expiration = this.getExpiration(topic);
                return typeof expiration !== "undefined";
            }
            catch (e) {
                return false;
            }
        };
        this.set = (topic, expiration) => {
            this.isInitialized();
            this.expirations.set(topic, expiration);
            this.checkExpiry(topic, expiration);
            this.events.emit(constants_1.EXPIRER_EVENTS.created, {
                topic,
                expiration,
            });
        };
        this.get = topic => {
            this.isInitialized();
            return this.getExpiration(topic);
        };
        this.del = (topic) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.isInitialized();
            const exists = yield this.has(topic);
            if (exists) {
                const expiration = this.getExpiration(topic);
                this.expirations.delete(topic);
                this.events.emit(constants_1.EXPIRER_EVENTS.deleted, {
                    topic,
                    expiration,
                });
            }
        });
        this.on = (event, listener) => {
            this.events.on(event, listener);
        };
        this.once = (event, listener) => {
            this.events.once(event, listener);
        };
        this.off = (event, listener) => {
            this.events.off(event, listener);
        };
        this.removeListener = (event, listener) => {
            this.events.removeListener(event, listener);
        };
        this.core;
        this.logger = logger_1.generateChildLogger(logger, this.name);
    }
    get context() {
        return logger_1.getLoggerContext(this.logger);
    }
    get storageKey() {
        return this.storagePrefix + this.version + "//" + this.name;
    }
    get length() {
        return this.expirations.size;
    }
    get topics() {
        return Array.from(this.expirations.keys());
    }
    get values() {
        return Array.from(this.expirations.values());
    }
    setExpirations(expirations) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.core.storage.setItem(this.storageKey, expirations);
        });
    }
    getExpirations() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const expirations = yield this.core.storage.getItem(this.storageKey);
            return expirations;
        });
    }
    persist() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.setExpirations(this.values);
            this.events.emit(constants_1.EXPIRER_EVENTS.sync);
        });
    }
    restore() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                const persisted = yield this.getExpirations();
                if (typeof persisted === "undefined")
                    return;
                if (!persisted.length)
                    return;
                if (this.expirations.size) {
                    const error = utils_1.ERROR.RESTORE_WILL_OVERRIDE.format({
                        context: this.name,
                    });
                    this.logger.error(error.message);
                    throw new Error(error.message);
                }
                this.cached = persisted;
                this.logger.debug(`Successfully Restored expirations for ${this.name}`);
                this.logger.trace({ type: "method", method: "restore", expirations: this.values });
            }
            catch (e) {
                this.logger.debug(`Failed to Restore expirations for ${this.name}`);
                this.logger.error(e);
            }
        });
    }
    getExpiration(topic) {
        const expiration = this.expirations.get(topic);
        if (!expiration) {
            const error = utils_1.ERROR.NO_MATCHING_ID.format({
                context: this.name,
                topic,
            });
            throw new Error(error.message);
        }
        return expiration;
    }
    checkExpiry(topic, expiration) {
        const { expiry } = expiration;
        const msToTimeout = time_1.toMiliseconds(expiry) - Date.now();
        if (msToTimeout <= 0)
            this.expire(topic, expiration);
    }
    expire(topic, expiration) {
        this.expirations.delete(topic);
        this.events.emit(constants_1.EXPIRER_EVENTS.expired, {
            topic,
            expiration,
        });
    }
    checkExpirations() {
        this.expirations.forEach((expiration, topic) => this.checkExpiry(topic, expiration));
    }
    registerEventListeners() {
        this.core.heartbeat.on(heartbeat_1.HEARTBEAT_EVENTS.pulse, () => this.checkExpirations());
        this.events.on(constants_1.EXPIRER_EVENTS.created, (createdEvent) => {
            const eventName = constants_1.EXPIRER_EVENTS.created;
            this.logger.info(`Emitting ${eventName}`);
            this.logger.debug({ type: "event", event: eventName, data: createdEvent });
            this.persist();
        });
        this.events.on(constants_1.EXPIRER_EVENTS.expired, (expiredEvent) => {
            const eventName = constants_1.EXPIRER_EVENTS.expired;
            this.logger.info(`Emitting ${eventName}`);
            this.logger.debug({ type: "event", event: eventName, data: expiredEvent });
            this.persist();
        });
        this.events.on(constants_1.EXPIRER_EVENTS.deleted, (deletedEvent) => {
            const eventName = constants_1.EXPIRER_EVENTS.deleted;
            this.logger.info(`Emitting ${eventName}`);
            this.logger.debug({ type: "event", event: eventName, data: deletedEvent });
            this.persist();
        });
    }
    isInitialized() {
        if (!this.initialized) {
            throw new Error(utils_1.ERROR.NOT_INITIALIZED.stringify(this.name));
        }
    }
}
exports.Expirer = Expirer;
