// @walletconnect/client/dist/cjs/constants/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLIENT_STORAGE_OPTIONS = exports.CLIENT_EVENTS = exports.CLIENT_SHORT_TIMEOUT = exports.CLIENT_DEFAULT = exports.CLIENT_STORAGE_PREFIX = exports.CLIENT_CONTEXT = exports.CLIENT_VERSION = exports.CLIENT_PROTOCOL = void 0;
exports.CLIENT_PROTOCOL = "wc";
exports.CLIENT_VERSION = 2;
exports.CLIENT_CONTEXT = "client";
exports.CLIENT_STORAGE_PREFIX = `${exports.CLIENT_PROTOCOL}@${exports.CLIENT_VERSION}:${exports.CLIENT_CONTEXT}:`;
exports.CLIENT_DEFAULT = {
    name: exports.CLIENT_CONTEXT,
    logger: "error",
    controller: false,
    relayUrl: "wss://relay.walletconnect.com",
};
exports.CLIENT_SHORT_TIMEOUT = 50;
exports.CLIENT_EVENTS = {
    session_proposal: "session_proposal",
    session_update: "session_update",
    session_extend: "session_extend",
    session_ping: "session_ping",
    pairing_ping: "pairing_ping",
    session_delete: "session_delete",
    session_expire: "session_expire",
    pairing_delete: "pairing_delete",
    pairing_expire: "pairing_expire",
    session_request: "session_request",
    session_event: "session_event",
};
exports.CLIENT_STORAGE_OPTIONS = {
    database: ":memory:",
};
